File: src/core/cartridge.ts
/**
 * NES 卡帶模組
 * 
 * iNES 格式說明：
 * - Header (16 bytes)
 *   - Bytes 0-3: "NES" + 0x1A
 *   - Byte 4: PRG ROM 大小 (16KB 為單位)
 *   - Byte 5: CHR ROM 大小 (8KB 為單位)
 *   - Byte 6: Flags 6 (mapper 低 4 位元, 鏡像模式等)
 *   - Byte 7: Flags 7 (mapper 高 4 位元, NES 2.0 標識)
 *   - Bytes 8-15: 其他標誌
 * - Trainer (512 bytes, 可選)
 * - PRG ROM
 * - CHR ROM
 */

import { Mapper, createMapper } from '../mappers';

/**
 * 鏡像模式
 */
export enum MirrorMode {
  /** 水平鏡像 (垂直排列) */
  Horizontal = 0,
  /** 垂直鏡像 (水平排列) */
  Vertical = 1,
  /** 單螢幕 (低) */
  SingleScreenLow = 2,
  /** 單螢幕 (高) */
  SingleScreenHigh = 3,
  /** 四螢幕 */
  FourScreen = 4,
}

/**
 * 卡帶標頭資訊
 */
export interface CartridgeHeader {
  /** PRG ROM 大小 (位元組) */
  prgRomSize: number;
  /** CHR ROM 大小 (位元組) */
  chrRomSize: number;
  /** Mapper 編號 */
  mapperNumber: number;
  /** 鏡像模式 */
  mirrorMode: MirrorMode;
  /** 是否有電池備份 */
  hasBattery: boolean;
  /** 是否有 Trainer */
  hasTrainer: boolean;
  /** 是否為 NES 2.0 格式 */
  isNes2: boolean;
}

/**
 * 卡帶類別
 */
export class Cartridge {
  /** PRG ROM 資料 */
  private prgRom: Uint8Array = new Uint8Array(0);
  
  /** CHR ROM 資料 (如果為 0 則使用 CHR RAM) */
  private chrRom: Uint8Array = new Uint8Array(0);
  
  /** PRG RAM (8KB) */
  private prgRam: Uint8Array = new Uint8Array(8192);
  
  /** CHR RAM (如果卡帶無 CHR ROM) */
  private chrRam: Uint8Array = new Uint8Array(8192);
  
  /** 是否使用 CHR RAM */
  private usesChrRam: boolean = false;
  
  /** 卡帶標頭資訊 */
  private header: CartridgeHeader | null = null;
  
  /** Mapper */
  private mapper: Mapper | null = null;
  
  /** 鏡像模式 */
  private mirrorMode: MirrorMode = MirrorMode.Horizontal;

  constructor() {
    this.prgRam.fill(0);
    this.chrRam.fill(0);
  }

  /**
   * 從 ArrayBuffer 載入 ROM
   */
  public loadRom(data: ArrayBuffer): boolean {
    const bytes = new Uint8Array(data);
    
    // 驗證 iNES 標頭
    if (bytes[0] !== 0x4E || bytes[1] !== 0x45 ||
        bytes[2] !== 0x53 || bytes[3] !== 0x1A) {
      console.error('無效的 iNES 標頭');
      return false;
    }

    // 解析標頭
    const prgRomBanks = bytes[4];
    const chrRomBanks = bytes[5];
    const flags6 = bytes[6];
    const flags7 = bytes[7];

    this.header = {
      prgRomSize: prgRomBanks * 16384,
      chrRomSize: chrRomBanks * 8192,
      mapperNumber: ((flags6 >> 4) & 0x0F) | (flags7 & 0xF0),
      mirrorMode: (flags6 & 0x01) ? MirrorMode.Vertical : MirrorMode.Horizontal,
      hasBattery: (flags6 & 0x02) !== 0,
      hasTrainer: (flags6 & 0x04) !== 0,
      isNes2: (flags7 & 0x0C) === 0x08,
    };

    // 四螢幕模式覆蓋
    if (flags6 & 0x08) {
      this.header.mirrorMode = MirrorMode.FourScreen;
    }

    this.mirrorMode = this.header.mirrorMode;

    // 計算 ROM 資料偏移
    let offset = 16;
    if (this.header.hasTrainer) {
      offset += 512;
    }

    // 載入 PRG ROM
    this.prgRom = new Uint8Array(this.header.prgRomSize);
    this.prgRom.set(bytes.slice(offset, offset + this.header.prgRomSize));
    offset += this.header.prgRomSize;

    // 載入 CHR ROM 或使用 CHR RAM
    if (this.header.chrRomSize > 0) {
      this.chrRom = new Uint8Array(this.header.chrRomSize);
      this.chrRom.set(bytes.slice(offset, offset + this.header.chrRomSize));
      this.usesChrRam = false;
    } else {
      // 使用 CHR RAM
      this.chrRam = new Uint8Array(8192);
      this.chrRam.fill(0);
      this.usesChrRam = true;
    }

    // 建立 Mapper
    this.mapper = createMapper(
      this.header.mapperNumber,
      prgRomBanks,
      chrRomBanks
    );

    if (!this.mapper) {
      console.error(`不支援的 Mapper: ${this.header.mapperNumber}`);
      return false;
    }

    console.log(`ROM 載入成功:`);
    console.log(`  PRG ROM: ${this.header.prgRomSize / 1024}KB`);
    console.log(`  CHR ROM: ${this.header.chrRomSize / 1024}KB`);
    console.log(`  Mapper: ${this.header.mapperNumber}`);
    console.log(`  鏡像: ${MirrorMode[this.header.mirrorMode]}`);

    return true;
  }

  /**
   * CPU 讀取
   */
  public cpuRead(address: number): number {
    // PRG RAM ($6000-$7FFF)
    if (address >= 0x6000 && address < 0x8000) {
      return this.prgRam[address & 0x1FFF];
    }

    // PRG ROM ($8000-$FFFF)
    if (address >= 0x8000 && this.mapper) {
      const mappedAddr = this.mapper.cpuMapRead(address);
      if (mappedAddr !== null) {
        return this.prgRom[mappedAddr % this.prgRom.length];
      }
    }

    return 0;
  }

  /**
   * CPU 寫入
   */
  public cpuWrite(address: number, data: number): void {
    // PRG RAM ($6000-$7FFF)
    if (address >= 0x6000 && address < 0x8000) {
      this.prgRam[address & 0x1FFF] = data;
      return;
    }

    // Mapper 暫存器寫入
    if (address >= 0x8000 && this.mapper) {
      const result = this.mapper.cpuMapWrite(address, data);
      
      // 檢查鏡像模式是否改變
      if (result?.mirrorMode !== undefined) {
        this.mirrorMode = result.mirrorMode;
      }
    }
  }

  /**
   * PPU 讀取
   */
  public ppuRead(address: number): number {
    if (address < 0x2000) {
      if (this.mapper) {
        const mappedAddr = this.mapper.ppuMapRead(address);
        if (mappedAddr !== null) {
          if (this.usesChrRam) {
            return this.chrRam[mappedAddr % this.chrRam.length];
          } else {
            return this.chrRom[mappedAddr % this.chrRom.length];
          }
        }
      }
      
      // 直接存取
      if (this.usesChrRam) {
        return this.chrRam[address];
      } else {
        return this.chrRom[address % this.chrRom.length];
      }
    }
    return 0;
  }

  /**
   * PPU 寫入
   */
  public ppuWrite(address: number, data: number): void {
    if (address < 0x2000) {
      if (this.mapper) {
        const mappedAddr = this.mapper.ppuMapWrite(address);
        if (mappedAddr !== null && this.usesChrRam) {
          this.chrRam[mappedAddr % this.chrRam.length] = data;
          return;
        }
      }
      
      // 直接寫入 CHR RAM
      if (this.usesChrRam) {
        this.chrRam[address] = data;
      }
    }
  }

  /**
   * 取得鏡像模式
   */
  public getMirrorMode(): MirrorMode {
    return this.mirrorMode;
  }

  /**
   * 取得卡帶標頭資訊
   */
  public getHeader(): CartridgeHeader | null {
    return this.header;
  }

  /**
   * 重置卡帶
   */
  public reset(): void {
    this.mapper?.reset?.();
  }

  /**
   * 掃描線計數 (用於 MMC3 等 Mapper)
   */
  public scanline(): void {
    this.mapper?.scanline?.();
  }

  /**
   * CPU 週期計數 (用於 Bandai FCG 等 Mapper)
   */
  public cpuClock(): void {
    this.mapper?.cpuClock?.();
  }

  /**
   * 檢查 Mapper IRQ
   */
  public checkIrq(): boolean {
    if (this.mapper && 'getIrqPending' in this.mapper) {
      return (this.mapper as any).getIrqPending();
    }
    return false;
  }

  // ===== 序列化 (存檔) =====

  /** 儲存狀態 */
  public saveState(): object {
    return {
      prgRam: Array.from(this.prgRam),
      chrRam: this.usesChrRam ? Array.from(this.chrRam) : [],
      mirrorMode: this.mirrorMode,
    };
  }

  /** 載入狀態 */
  public loadState(state: any): void {
    this.prgRam.set(state.prgRam);
    if (this.usesChrRam && state.chrRam.length > 0) {
      this.chrRam.set(state.chrRam);
    }
    this.mirrorMode = state.mirrorMode;
  }
}

File: nes-wasm/src/cartridge.rs
// ============================================================
// NES 卡帶模擬 - iNES 格式解析與記憶體管理
// ============================================================
// 負責解析 iNES 和 NES 2.0 格式的 ROM 檔案，
// 並管理 PRG ROM、CHR ROM/RAM 的存取。
//
// iNES 格式：
// - 16 位元組標頭
// - 可選的 512 位元組訓練器（Trainer）
// - PRG ROM 資料（16KB 為單位）
// - CHR ROM 資料（8KB 為單位，可能為 0）
//
// 參考：https://www.nesdev.org/wiki/INES
// ============================================================

use crate::ppu::MirrorMode;
use crate::mappers::*;

/// iNES 標頭結構
pub struct CartridgeHeader {
    /// PRG ROM 大小（16KB 為單位）
    pub prg_rom_banks: u8,
    /// CHR ROM 大小（8KB 為單位，0 表示使用 CHR RAM）
    pub chr_rom_banks: u8,
    /// Mapper 編號
    pub mapper_id: u8,
    /// 鏡像模式
    pub mirror_mode: MirrorMode,
    /// 是否有電池供電的 SRAM
    pub has_battery: bool,
    /// 是否有訓練器資料
    pub has_trainer: bool,
}

/// NES 卡帶
pub struct Cartridge {
    /// 卡帶標頭資訊
    pub header: CartridgeHeader,
    /// PRG ROM 資料
    pub prg_rom: Vec<u8>,
    /// CHR ROM/RAM 資料
    pub chr_data: Vec<u8>,
    /// PRG RAM（8KB，可能有電池供電）
    pub prg_ram: Vec<u8>,
    /// 是否使用 CHR RAM
    pub chr_ram: bool,
    /// Mapper 實例
    pub mapper: Box<dyn MapperTrait>,
    /// 是否已載入 ROM
    pub loaded: bool,
}

impl Cartridge {
    /// 建立空的卡帶
    pub fn new() -> Self {
        Cartridge {
            header: CartridgeHeader {
                prg_rom_banks: 0,
                chr_rom_banks: 0,
                mapper_id: 0,
                mirror_mode: MirrorMode::Horizontal,
                has_battery: false,
                has_trainer: false,
            },
            prg_rom: Vec::new(),
            chr_data: Vec::new(),
            prg_ram: vec![0; 8192], // 8KB PRG RAM
            chr_ram: false,
            mapper: Box::new(Mapper0::new(1, 1)),
            loaded: false,
        }
    }

    /// 載入 ROM 資料
    pub fn load_rom(&mut self, data: &[u8]) -> bool {
        // 檢查 iNES 標頭魔數 "NES\x1A"
        if data.len() < 16 || data[0] != 0x4E || data[1] != 0x45 ||
           data[2] != 0x53 || data[3] != 0x1A {
            return false;
        }

        // 解析標頭
        let prg_banks = data[4];
        let chr_banks = data[5];
        let flags6 = data[6];
        let flags7 = data[7];

        // Mapper 編號（低 4 位元在 flags6，高 4 位元在 flags7）
        let mapper_id = (flags7 & 0xF0) | (flags6 >> 4);

        // 鏡像模式
        let mirror_mode = if flags6 & 0x08 != 0 {
            MirrorMode::FourScreen
        } else if flags6 & 0x01 != 0 {
            MirrorMode::Vertical
        } else {
            MirrorMode::Horizontal
        };

        let has_battery = flags6 & 0x02 != 0;
        let has_trainer = flags6 & 0x04 != 0;

        self.header = CartridgeHeader {
            prg_rom_banks: prg_banks,
            chr_rom_banks: chr_banks,
            mapper_id,
            mirror_mode,
            has_battery,
            has_trainer,
        };

        // 計算資料偏移
        let mut offset = 16;
        if has_trainer {
            offset += 512; // 跳過訓練器
        }

        // 讀取 PRG ROM
        let prg_size = prg_banks as usize * 16384; // 16KB per bank
        if offset + prg_size > data.len() {
            return false;
        }
        self.prg_rom = data[offset..offset + prg_size].to_vec();
        offset += prg_size;

        // 讀取 CHR ROM/RAM
        if chr_banks > 0 {
            let chr_size = chr_banks as usize * 8192; // 8KB per bank
            if offset + chr_size > data.len() {
                // 某些 ROM 的 CHR 資料可能不完整，用 0 填充
                self.chr_data = vec![0; chr_size];
                let available = data.len().saturating_sub(offset);
                if available > 0 {
                    self.chr_data[..available].copy_from_slice(&data[offset..offset + available]);
                }
            } else {
                self.chr_data = data[offset..offset + chr_size].to_vec();
            }
            self.chr_ram = false;
        } else {
            // 使用 CHR RAM（8KB）
            self.chr_data = vec![0; 8192];
            self.chr_ram = true;
        }

        // 重置 PRG RAM
        self.prg_ram = vec![0; 8192];

        // 建立 Mapper
        self.mapper = create_mapper(mapper_id, prg_banks, chr_banks);

        // Mapper 253 (Waixing VRC4) 需要額外的 CHR RAM 空間
        // 在 CHR ROM 末尾追加 8KB CHR RAM，用於動態 CHR bank 替換
        if mapper_id == 253 && !self.chr_ram {
            let chr_rom_size = self.chr_data.len();
            self.chr_data.resize(chr_rom_size + 8192, 0);
        }

        self.loaded = true;

        true
    }

    /// 重置卡帶
    pub fn reset(&mut self) {
        self.mapper.reset();
    }

    /// CPU 讀取
    pub fn cpu_read(&self, addr: u16) -> u8 {
        // PRG RAM ($6000-$7FFF) — 直接存取，不依賴 Mapper
        if addr >= 0x6000 && addr < 0x8000 {
            let index = (addr - 0x6000) as usize;
            return self.prg_ram.get(index).copied().unwrap_or(0);
        }

        if let Some(mapped) = self.mapper.cpu_read(addr) {
            if addr >= 0x8000 {
                // PRG ROM
                let index = mapped as usize % self.prg_rom.len().max(1);
                self.prg_rom.get(index).copied().unwrap_or(0)
            } else {
                0
            }
        } else {
            0
        }
    }

    /// CPU 寫入
    pub fn cpu_write(&mut self, addr: u16, data: u8) {
        if addr >= 0x6000 && addr < 0x8000 {
            // PRG RAM 寫入
            let index = (addr - 0x6000) as usize;
            if index < self.prg_ram.len() {
                self.prg_ram[index] = data;
            }
        }

        // 通知 Mapper（可能觸發 bank 切換等）
        if let Some(result) = self.mapper.cpu_write(addr, data) {
            if let Some(mode) = result.mirror_mode {
                self.header.mirror_mode = mode;
            }
        }
    }

    /// PPU 讀取（CHR ROM/RAM）
    pub fn ppu_read(&self, addr: u16) -> u8 {
        if let Some(mapped) = self.mapper.ppu_read(addr) {
            let index = mapped as usize % self.chr_data.len().max(1);
            self.chr_data.get(index).copied().unwrap_or(0)
        } else {
            0
        }
    }

    /// PPU 寫入（只有 CHR RAM 可寫）
    pub fn ppu_write(&mut self, addr: u16, data: u8) {
        if self.chr_ram {
            if let Some(mapped) = self.mapper.ppu_write(addr) {
                let index = mapped as usize;
                if index < self.chr_data.len() {
                    self.chr_data[index] = data;
                }
            }
        }
    }

    /// 通知 Mapper 掃描線計數（用於 MMC3 等）
    pub fn scanline(&mut self) {
        self.mapper.scanline();
    }

    /// 通知 Mapper CPU 週期計數
    pub fn cpu_clock(&mut self) {
        self.mapper.cpu_clock();
    }

    /// 檢查 Mapper IRQ
    pub fn check_irq(&mut self) -> bool {
        self.mapper.check_irq()
    }

    /// 取得目前的鏡像模式
    pub fn mirror_mode(&self) -> MirrorMode {
        self.header.mirror_mode
    }
}
